letsencrypt/acme/messages.py

```diff
@@ -0,0 +1,491 @@
+
+    def _fields_to_json(self):
+        fields = {}
+        if self.recovery_token is not None:
+            fields["recoveryToken"] = self.recovery_token
+        if self.identifier is not None:
+            fields["identifier"] = self.identifier
+        if self.jwk is not None:
+            fields["jwk"] = self.jwk
+        return fields
+
+    @classmethod
+    def _from_valid_json(cls, jobj):
+        jwk = jobj.get("jwk")
+        if jwk is not None:
+            jwk = jose.JWK.from_json(jwk, validate=False)
+        return cls(recovery_token=jobj.get("recoveryToken"),
+                   identifier=jobj.get("identifier"), jwk=jwk)
+
+
+@Message.register
+class AuthorizationRequest(Message):
+    """ACME "authorizationRequest" message.
+
+    :ivar str session_id: "sessionID" from the server challenge
+    :ivar str nonce: Nonce from the server challenge
+    :ivar list responses: List of completed challenges
+    :ivar signature: Signature (:class:`letsencrypt.acme.other.Signature`).
+    :ivar contact: TODO
+
+    """
+    acme_type = "authorizationRequest"
+    schema = util.load_schema(acme_type)
+    __slots__ = ("session_id", "nonce", "responses", "signature", "contact")
+
+    @classmethod
+    def create(cls, name, key, sig_nonce=None, **kwargs):
+        """Create signed "authorizationRequest".
+
+        :param str name: Hostname
+
+        :param key: Key used for signing.
+        :type key: :class:`Crypto.PublicKey.RSA`
+
+        :param str sig_nonce: Nonce used for signature. Useful for testing.
+
+        :returns: Signed "authorizationRequest" ACME message.
+        :rtype: :class:`AuthorizationRequest`
+
+        """
+        # pylint: disable=too-many-arguments
+        signature = other.Signature.from_msg(
+            name + kwargs["nonce"], key, sig_nonce)
+        return cls(
+            signature=signature, contact=kwargs.pop("contact", []), **kwargs)
+
+    def verify(self, name):
+        """Verify signature.
+
+        :param str name: Hostname
+
+        :returns: True iff ``signature`` can be verified, False otherwise.
+        :rtype: bool
+
+        """
+        # TODO: must also check that the public key encoded in the JWK object
+        # is the correct key for a given context.
+        return self.signature.verify(name + self.nonce)
+
+    def _fields_to_json(self):
+        fields = {
+            "sessionID": self.session_id,
+            "nonce": jose.b64encode(self.nonce),
+            "responses": self.responses,
+            "signature": self.signature,
+        }
+        if self.contact:
+            fields["contact"] = self.contact
+        return fields
+
+    @classmethod
+    def _from_valid_json(cls, jobj):
+        return cls(session_id=jobj["sessionID"],
+                   nonce=jose.b64decode(jobj["nonce"]),
+                   responses=jobj["responses"],
+                   signature=other.Signature.from_json(
+                       jobj["signature"], validate=False),
```

Comment:
Validation already happens in `Message.from_json`, which calls `_from_valid_json` only once the whole message (including all subjects, such as signatures or JWKs) is validated against ACME JSON schema. We will notice if someone tries hard to call `_from_valid_json` directly as it is private.

As mentioned in https://github.com/letsencrypt/lets-encrypt-preview/pull/233#discussion_r24710754 whole validation business is a bit changed in my recent local commits, so this will change anyway.

Therefore, please allow me to leave the code as is in this PR.